### ISSUE TITLE
cap_gstreamer::open uses CV_ERROR

### ISSUE THREAD

**Issue by TheExDeus** [opencv/opencv#7385]
##### System information (version)

- OpenCV => 3.1
- Operating System / Platform => Windows 10 64bit and Ubuntu Linux 64bit
- Compiler => gcc 5.3.0

##### Detailed description

The gstreamer implementation in videoio module uses CV_ERROR() on open() instead of just returning false like all other implementations. This causes an exception to be thrown, but in projects where exceptions are disabled we cannot have that - exceptions are just handled as crashes. But in this case an error can be thrown only if a file doesn't exist or if a pipeline is wrong. In our software we have a modular api that allows several pipelines in a priority order - if a pipeline cannot be used it goes to the next one. It doesn't work because the first pipeline crashes instead of returning false like the ffmpeg, dshow or any other implementation. This is clearly an inconsistency in OpenCV api.

##### Steps to reproduce
```.cpp
    int main() {
      cv::VideoCapture cap("shmsrc /tmp/socket ! video/x-raw, 1920, 1080, 1/60, format=UYVY ! videoconvert ! video/x-raw, BGR ! appsink sync=true"); 
    }
```

That will crash if the socket doesn't exist. It would be good if it returned false if used with Open() and just set internally as not-running, so we could check with isOpened.

**Comment by alalek** [comment]
> but in projects where exceptions are disabled

You can't disable C++ exceptions in OpenCV. Exception is not a crash, but unhandled exception is crash.

> the first pipeline crashes instead of returning false like the ffmpeg

There are several options available:
- build with disabled GStreamer
- use API with `apiPreference` argument to force ffmpeg
- apply fix #6359

Closing as duplicate of #6357

**Comment by TheExDeus** [comment]
But then why some backends throw the exception and some don't? At least be consistent. This is the only place in OpenCV where I see a recoverable error to throw an exception. In all other places it really is unrecoverable.
Like I have no problems that asserts in functions make the program crash. For example having a kernel with size 0 in cv::blur or giving an empty Mat to a function that shouldn't take it. These things I can manually check before hand. But I cannot check that a gstreamer pipeline will run before running it. And it shouldn't be a crash.
Disabling gstreamer is not an option when you actually are using gstreamer pipelines. The fix here is simple - remove CV_ERROR from gstreamer backend open() functions just like they are not present in all other backends. I can make a pull request if required. #6359 is a bad way to fix it considering that only gstreamer backend can call an exception - the rest just return false - thus the rest of the backends are not actually tried.
For #6357 I propose making a simple list of all the available backends on runtime and try one after another just like you did, but without checking for exception (as that won't work anyway) and instead check for isOpened().
